Fix escaping of list items from triplets

Deadpikle opened the pull request:
This PR fixes some escaping issues of list items that could come from triplets. e.g., if you set multiple architectures for VCPKG_OSX_ARCHITECTURES, the arguments were not escaped properly. This PR fixes that.

The code from this PR was pulled from #12657 per the discussions there and was done entirely by @strega-nil.

- What does your PR fix?

No closed issues.

- Which triplets are supported/not supported? Have you updated the CI baseline?

No updates to CI. This doesn't directly affect any triplets, but in theory, it will allow you to set multiple architectures (e.g. for a macOS build) in a triplet.

Deadpikle committed:
Fix escaping of list items from triplets
e.g., if you set multiple architectures for VCPKG_OSX_ARCHITECTURES, the arguments were not escaped properly. This PR fixes that.
The code from this PR was pulled from #12657 and was done entirely by @strega-nil.

Neumann-A commented:
Wondering if usage of
```
 cmake_parse_arguments(PARSE_ARGV <N> <prefix> <options>
                      <one_value_keywords> <multi_value_keywords>)
```
would be better within VCPKG

strega-nil commented:
@Neumann-A seems reasonable

Neumann-A commented on scripts/cmake/vcpkg_execute_required_process.cmake:
```diff
@@ -34,7 +34,9 @@
 function(vcpkg_execute_required_process)
-    cmake_parse_arguments(vcpkg_execute_required_process "ALLOW_IN_DOWNLOAD_MODE" "WORKING_DIRECTORY;LOGNAME" "COMMAND" ${ARGN})
+    # cmake_parse_arguments ignores one level of escaping, so make that two levels of escaping
+    string(REPLACE "\\;" "\\\\;" _ver_ESCAPED_ARGN "${ARGN}")
+    cmake_parse_arguments(vcpkg_execute_required_process "ALLOW_IN_DOWNLOAD_MODE" "WORKING_DIRECTORY;LOGNAME" "COMMAND" "${_ver_ESCAPED_ARGN}")
```
Try
```
 cmake_parse_arguments(PARSE_ARGV <N> <prefix> <options>
                      <one_value_keywords> <multi_value_keywords>)
```
instead of escaping everything
```suggestion
     cmake_parse_arguments(PARSE_ARGV 0 vcpkg_execute_required_process "ALLOW_IN_DOWNLOAD_MODE" "WORKING_DIRECTORY;LOGNAME" "COMMAND")
```

Neumann-A commented on scripts/cmake/vcpkg_configure_cmake.cmake:
```diff
@@ -211,6 +211,8 @@ function(vcpkg_configure_cmake)
     endif()
 
+    # make sure to escape multiple architectures in VCPKG_TARGET_ARCHITECTURE
+    string(REPLACE ";" "\\;" _VCPKG_TARGET_ARCHITECTURE "${VCPKG_TARGET_ARCHITECTURE}")
```
Try using 
```
 cmake_parse_arguments(PARSE_ARGV <N> <prefix> <options>
                      <one_value_keywords> <multi_value_keywords>)

```
at the top instead of escaping

JackBoosY commented:
pbc regression will be fixed in #12732

Deadpikle committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into fix/list-escaping

Deadpikle committed:
Replace some escaping with cmake_parse_arguments

Deadpikle committed:
Fix ${_bn_PRERUN_SHELL_DEBUG} having quotes

The other vcpkg_execute_required_process in build_nmake didn't have quotes,
and now the quotes don't seem to be needed due to changes in
vcpkg_execute_required_process.

Deadpikle committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into fix/list-escaping

Deadpikle committed:
Remove some unneeded escaping in aws-sdk-cpp

Deadpikle committed:
Use GitHub for soil lib rather than URL

http://www.lonesock.net/files/soil.zip is down as of writing.
Use closest GitHub mirror we could find.

Deadpikle committed:
Remove some unneeded quote escaping in gdal

Deadpikle committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into fix/list-escaping

JackBoosY commented:
Pinging @Deadpikle  for response. Is work still being done for this PR?

Deadpikle commented:
@JackBoosY Hi. Sorry for dropping off the face of the planet in this PR. I had to do some other things for work, and I haven't been able to get back to this. **I don't think I'll be able to get back to this in the near future**, so either someone else needs to pick up this PR and its associated fixes (are they even necessary with other work that's been done since August? I haven't been following.) or this PR needs to be closed.

Apologies for not finishing it up. I was having trouble with the CI failing on something different every time.

JackBoosY commented:
@Deadpikle That's okay, please reopen this PR when you return to work.
